From daniel.sangorrin@toshiba.co.jp Mon Jan 20 22:23:22 2014
From: Daniel Sangorrin <daniel.sangorrin@toshiba.co.jp>
Date: Tue, 21 Jan 2014 15:23:06 +0900
Subject: [PATCH 1/2] Xilinx: ARM: BSP: prevent DMA into lower memory
To: ltsi-dev@lists.linuxfoundation.org
Cc: gregkh@linuxfoundation.org, michal.simek@xilinx.com
Message-ID: <1390285387-30514-3-git-send-email-daniel.sangorrin@toshiba.co.jp>


From: John Linn <john.linn@xilinx.com>

The DMA zone from 2.6.39 is no longer supported such that
a new method was needed. The old method was lost in the
move to 3.0 and USB was seeing failures.
(commit 83e198c01c381a1d90ba07e241a517d1dabf7c84 in Xilinx
repository)

Signed-off-by: Daniel Sangorrin <daniel.sangorrin@toshiba.co.jp>
Signed-off-by: Yoshitake Kobayashi <yoshitake.kobayashi@toshiba.co.jp>
---
 arch/arm/mach-zynq/common.c |   19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

--- a/arch/arm/mach-zynq/common.c
+++ b/arch/arm/mach-zynq/common.c
@@ -25,7 +25,7 @@
 #include <linux/of_irq.h>
 #include <linux/of_platform.h>
 #include <linux/of.h>
-
+#include <linux/memblock.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/map.h>
 #include <asm/mach/time.h>
@@ -39,6 +39,22 @@
 
 void __iomem *zynq_scu_base;
 
+/**
+ * zynq_memory_init() - Initialize special memory
+ *
+ * We need to stop things allocating the low memory as DMA can't work in
+ * the 1st 512K of memory.  Using reserve vs remove is not totally clear yet.
+ */
+static void __init zynq_memory_init(void)
+{
+	/*
+	 * Reserve the 0-0x4000 addresses (before page tables and kernel)
+	 * which can't be used for DMA
+	 */
+	if (!__pa(PAGE_OFFSET))
+		memblock_reserve(0, 0x4000);
+}
+
 static struct of_device_id zynq_of_bus_ids[] __initdata = {
 	{ .compatible = "simple-bus", },
 	{}
@@ -113,5 +129,6 @@ DT_MACHINE_START(XILINX_EP107, "Xilinx Z
 	.init_machine	= zynq_init_machine,
 	.init_time	= zynq_timer_init,
 	.dt_compat	= zynq_dt_match,
+	.reserve	= zynq_memory_init,
 	.restart	= zynq_system_reset,
 MACHINE_END
